artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/converter/coreWrapper/CoreMessageWrapper.java

```diff
@@ -137,247 +175,187 @@ public final String getJMSCorrelationID() throws JMSException {
-      return ActiveMQDestination.createDestination(message.getRoutingType(), message.getAddressSimpleString());
-   }
-
-   @Override
-   public final void setJMSDestination(Destination destination) throws JMSException {
-      if (destination == null) {
-         message.setAddress((SimpleString) null);
-      } else {
-         message.setAddress(((ActiveMQDestination) destination).getSimpleAddress());
+   public SimpleString getDestination()  {
+      if (message.getAddress() == null || message.getAddress().isEmpty()) {
+         return null;
       }
-
+      return SimpleString.toSimpleString(AMQPMessageSupport.destination(message.getRoutingType(), message.getAddress()));
    }
 
-   @Override
-   public final int getJMSDeliveryMode() throws JMSException {
-      return message.isDurable() ? DeliveryMode.PERSISTENT : DeliveryMode.NON_PERSISTENT;
+   public final void setDestination(String destination)  {
+      message.setAddress(destination);
    }
 
-   @Override
-   public final void setJMSDeliveryMode(int deliveryMode) throws JMSException {
-      if (deliveryMode == DeliveryMode.PERSISTENT) {
-         message.setDurable(true);
-      } else if (deliveryMode == DeliveryMode.NON_PERSISTENT) {
-         message.setDurable(false);
-      } else {
-         throw new JMSException("Invalid mode " + deliveryMode);
-      }
+   public final int getJMSDeliveryMode()  {
+      return message.isDurable() ? PERSISTENT : NON_PERSISTENT;
    }
 
-   @Override
-   public final boolean getJMSRedelivered() throws JMSException {
-      return false;
+   public final void setDeliveryMode(int deliveryMode) throws ConversionException {
+      switch (deliveryMode) {
+         case PERSISTENT:
+            message.setDurable(true);
+            break;
+         case NON_PERSISTENT:
+            message.setDurable(false);
+            break;
+         default:
+            // There shouldn't be any other values used
+            throw new ConversionException("Invalid mode " + deliveryMode);
+      }
    }
 
-   @Override
-   public final void setJMSRedelivered(boolean redelivered) throws JMSException {
-      // no op
-   }
 
-   @Override
-   public final String getJMSType() throws JMSException {
+   public final String getJMSType()  {
       return MessageUtil.getJMSType(message);
    }
 
-   @Override
-   public final void setJMSType(String type) throws JMSException {
+   public final void setJMSType(String type)  {
       MessageUtil.setJMSType(message, type);
    }
 
-   @Override
-   public final long getJMSExpiration() throws JMSException {
+   public final long getExpiration()  {
       return message.getExpiration();
    }
 
-   @Override
-   public final void setJMSExpiration(long expiration) throws JMSException {
+   public final void setJMSExpiration(long expiration)  {
       message.setExpiration(expiration);
    }
 
-   @Override
-   public final long getJMSDeliveryTime() throws JMSException {
```

Comment:
Have we a possible bug here in use case of delivery delay in jms2.0 and message gets converted from amqp to core / vice versa.

Reply:
TTL is handled directly at the converter. I think this method was not being used.

Reply:
I was talking about DeliveryDelay (feature in jms 2.0) not TimeToLive

Reply:
getJMSDeliveryTime
long getJMSDeliveryTime()
                 throws JMSException
Gets the message's delivery time value.
When a message is sent, the JMSDeliveryTime header field is left unassigned. After completion of the send or publish method, it holds the delivery time of the message. This is the the difference, measured in milliseconds, between the delivery time and midnight, January 1, 1970 UTC.

A message's delivery time is the earliest time when a JMS provider may deliver the message to a consumer. The provider must not deliver messages before the delivery time has been reached.

Returns:
the message's delivery time value
Throws:
JMSException - if the JMS provider fails to get the delivery time due to some internal error.
Since:
JMS 2.0
See Also:
setJMSDeliveryTime(long)

Reply:
This is handled by the converter directly. The "JMS" facade was not using it in any way.

There are open wire scheduled delivery tests for this already.


On AMQP to Core Converter:

https://github.com/apache/activemq-artemis/blob/master/artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/converter/AmqpCoreConverter.java#L279-L300

On Core to AMQP Converter, the Converter will remove the Scheduled Delivery on Purpose, as part of:

https://github.com/apache/activemq-artemis/blob/master/artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/converter/CoreAmqpConverter.java#L300-L302



The only reason Actually AMQP To Core converter is treating scheduling is because we used to always convert to Core. On the recent versions of the broker the message will stay on its original format until the actuall delivery occurred.

If there is a bug it would not been cause by this change anyways but I think we are safe here.

Reply:
I will add some testing as a separate task from this. There's one possibility actually, when you send Core Message with a scheduled delivery through the AMQP Mirror the delivery delay may been stripped on that case. This would be a side effect of what you mentioned. I will test it and fix it if it's an issue. (not related to this change but I will validate it).

Reply:
Totally agree it's not related to this change. E.g. if its a bug was preexisting, just that on reviewing this change in detail it flagged to me this looks like a bug we may not know about, better to discuss it now than someone hit it in prod.

Reply:
@michaelandrepearce PR coming soon: https://issues.apache.org/jira/browse/ARTEMIS-3116

It was an issue indeed. I believe you would need AMQP Mirror to hit the issue.

PR coming soon.